tests/datastructures/test_dsbsparse.py

```diff
@@ -54,6 +54,28 @@ def test_zeros_like(
         assert (zeros.to_dense() == 0).all()
 
 
+def _create_new_block_sizes(
+    block_sizes: xp.ndarray, block_change_factor: float
+) -> xp.ndarray:
+    """Creates new block sizes based on the block change factor."""
+    # Create new block sizes.
+    rest = 0
+    updated_block_sizes = xp.array([], dtype=int)
+    for bs in block_sizes:
+        if sum(updated_block_sizes) < sum(block_sizes):
+            el = int(bs * block_change_factor) if bs * block_change_factor > 1 else 1
+            reps, rest = divmod(bs + rest, el) if bs + rest > el else (1, 0)
+            updated_block_sizes = xp.concatenate(
+                (updated_block_sizes, xp.ones(int(reps), dtype=int) * el)
+            )
+        else:
+            updated_block_sizes[-1] -= sum(updated_block_sizes) - sum(block_sizes)
+            break
+    if sum(updated_block_sizes) != sum(block_sizes):
+        updated_block_sizes[-1] += sum(block_sizes) - sum(updated_block_sizes)
+    return updated_block_sizes
+
+
```

Comment:
Not sure how you made this work but returns what i would expect :+1:. This seems to have a very similar flavor to the `get_section_sizes` method in the `mpi_utils`, could you check if you could clean this up a little bit along those lines using just Python lists instead of concatenating arrays? Also
```python
el = int(bs * block_change_factor) if bs * block_change_factor > 1 else 1
```
could just be 
```python
el = max(int(bs * block_change_factor), 1)
```
etc.

(since this only appears in the tests i don't really care if this stays like it is :smile:)

Reply:
It added some comments and your suggested changes. Hopefully it is a bit clearer now